[CUDA][HIP] warn incompatible redeclare

yxsamliu opened the pull request:
nvcc warns about the following code:

`void f();
__device__ void f() {}`

but clang does not since clang allows device function to overload host function.

Users want clang to emit similar warning to help code to be compatible with nvcc.

Since this may cause regression with existing code, the warning is off by default and can be enabled by -Wnvcc-compat.

It won't cause warning in system headers, even with -Wnvcc-compat.

yxsamliu commented:
ping

Artem-B commented on clang/test/SemaCUDA/function-redclare.cu:
```diff
@@ -0,0 +1,19 @@
+// expected-no-diagnostics
+#include "cuda.h"
+
+__device__ void f(); // redecl-note {{previous declaration is here}}
+
+void f() {} // redecl-warning {{incompatible host/device attribute with redeclaration: new declaration is __host__ function, old declaration is __device__ function. It will cause warning with nvcc}}
```
I'm not convinced that it's something we need to fix. As far as CUDA is concerned those are two different function overloads and it's business as usual with no warning required.

nvcc does not have a concept of attribute-based overloads, so for them a function with different attributes is a sign of a potential issue, but it's not the case for clang, IMO.

Artem-B replied:
I believe turning the warning on will never work in practice, because there will likely almost always be a ton of math functions with host and device overloads.
As I said, it is *very* common to have those and a lot of such overloads get pulled in from the headers pre-included by clang.

philnik777 replied:
IMO that looks like a nice "this is incompatible with X" warning like we have for GCC, different C/C++ versions etc., but I'm not an expert here, so maybe this isn't actually an incompatibility? FWIW neither the diagnostic nor you comment make this really clear to me. (Also I'd drop the `It will cause warning with nvcc` and make the flag something like `-Wnvcc-compat`)

Artem-B replied:
> "this is incompatible with X" warning like we have for GCC

CUDA compilation includes a lot of CUDA SDK and host headers. `<cmath>` is one of them. Compiler will *always* see a lot of implicitly `__host__` math functions declared/defined by the host headers and a lot of the same functions with `__device__` attribute declared/defined for the GPU. Whether or not user code has the same warning becomes rather moot, when *all* compilations will produce tons of these warnings. 

That said, we *may* be saved by the fact that (IIRC) compiler hides the warnings that come from the system headers.

It would be good to test it on a real CUDA compilation first. If we're guaranteed to have dozens of those warnings for every compilation, it would render the flag practically useless.

yxsamliu replied:
I checked with real HIP apps and the warnings will only show up for user's code or header files. There are no warnings for host/device redeclarations in HIP or clang headers since they are included as system headers.

yxsamliu committed:
[CUDA][HIP] warn incompatible redeclare

nvcc warns about the following code:

but clang does not since clang allows device function to
overload host function.

Users want clang to emit similar warning to help code to be
compatible with nvcc.

Since this may cause regression with existing code, the warning
is off by default and can be enabled by -Woffload-incompatible-redeclare.

It won't cause warning in system headers, even with
-Woffload-incompatible-redeclare.

yxsamliu commented:
ping

Our users keep requesting this feature since they want their HIP code works with both nvcc and clang. I tested it with real HIP apps and did not see warnings emitted for clang wrapper headers and HIP system headers. Only warnings for users' own code were emitted. Also since this warning is off by default, it won't affect normal users.

Artem-B reviewed (approved):
LGTM with some wording/namiung nits.

yxsamliu commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -9013,6 +9013,12 @@ def err_cuda_ovl_target : Error<
+  "incompatible host/device attribute with redeclaration: "
+  "new declaration is %select{__device__|__global__|__host__|__host__ __device__}0 function, "
+  "old declaration is %select{__device__|__global__|__host__|__host__ __device__}1 function. "
+  "It will cause warning with nvcc">,
```
will modify the diagnostic message

yxsamliu committed:
rename warning
